lib/rules/static-property-placement.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * @fileoverview Defines where React component static properties should be positioned.
+ * @author Daniel Mason
+ */
+'use strict';
+
+const Components = require('../util/Components');
+const docsUrl = require('../util/docsUrl');
+const astUtil = require('../util/ast');
+const propsUtil = require('../util/props');
+
+// ------------------------------------------------------------------------------
+// Positioning Options
+// ------------------------------------------------------------------------------
+const STATIC_PUBLIC_FIELD = 'static public field';
+const PROPERTY_ASSIGNMENT = 'property assignment';
+const POSITION_SETTINGS = [STATIC_PUBLIC_FIELD, PROPERTY_ASSIGNMENT];
+
+// ------------------------------------------------------------------------------
+// Rule messages
+// ------------------------------------------------------------------------------
+const STATIC_PUBLIC_FIELD_MESSAGE = '\'{{name}}\' should be declared as a static class property.';
+const PROPERTY_ASSIGNMENT_MESSAGE = '\'{{name}}\' should be declared outside the class body.';
+
+// ------------------------------------------------------------------------------
+// Properties to check
+// ------------------------------------------------------------------------------
+const propertiesToCheck = {
+  propTypes: propsUtil.isPropTypesDeclaration,
+  defaultProps: propsUtil.isDefaultPropsDeclaration,
+  childContextTypes: propsUtil.isChildContextTypesDeclaration,
+  contextTypes: propsUtil.isContextTypesDeclaration,
+  displayName: node => propsUtil.isDisplayNameDeclaration(astUtil.getPropertyNameNode(node))
+};
+
+const classProperties = Object.keys(propertiesToCheck);
+
+const schemaProperties = {};
+classProperties.map(property => {
+  schemaProperties[property] = {enum: POSITION_SETTINGS};
```

Comment:
Please don't use `.map` for side effects; instead of this, let's do:
```js
const fromEntries = require('object.fromentries');
const schemaProperties = fromEntries(classProperties.map(property => [property, {enum: POSITION_SETTINGS}]))l
```